File: frontend/src/lib/risk-check.ts
import { z } from "zod";

/** Public projection contract. No browser-side financial calculations. */
export const riskCheckSchema = z.object({
  portfolio_id: z.string(),
  result_id: z.string(),
  methodology_version: z.literal("risk-check-v1"),
  computed_at: z.string(),
  price_history_as_of: z.string().nullable(),
  status: z.enum(["ready", "limited"]),
  summary: z.string(),
  metrics: z.array(
    z.object({
      key: z.string(),
      label: z.string(),
      value: z.number().finite().nullable(),
      unit: z.enum(["usd", "fraction", "multiple", "days"]),
      horizon: z.string(),
      basis: z.string(),
      explanation: z.string(),
      source_field: z.string(),
    }),
  ),
  findings: z
    .array(
      z.object({
        key: z.string(),
        title: z.string(),
        severity: z.enum(["high", "elevated", "info"]),
        explanation: z.string(),
      }),
    )
    .max(3),
  limitations: z.array(z.string()),
  strategies: z
    .array(
      z.object({
        underlying: z.string(),
        expiry: z.string(),
        name: z.string(),
        leg_count: z.number(),
        premium_basis: z.enum([
          "entry",
          "current_mark",
          "mixed",
          "unavailable",
        ]),
        max_loss: z.number().finite().nullable(),
        max_gain: z.number().finite().nullable(),
        loss_status: z.enum(["bounded", "unbounded", "unavailable"]),
        gain_status: z.enum(["bounded", "unbounded", "unavailable"]),
      }),
    )
    .default([]),
});

export type RiskCheck = z.infer<typeof riskCheckSchema>;

/** Only explicit check requests trigger the heavier risk computation. */
export function isRiskCheckRequest(text: string): boolean {
  return /^(check (my|this) portfolio( risk)?|run (a |my )?risk check|检查我的(投资)?组合(风险)?|检查组合风险)[.!。！]?$/i.test(
    text.trim(),
  );
}

export function formatCheckValue(metric: RiskCheck["metrics"][number]): string {
  if (metric.value === null) return "Unavailable";
  switch (metric.unit) {
    case "usd":
      return new Intl.NumberFormat("en-US", {
        style: "currency",
        currency: "USD",
        maximumFractionDigits: 0,
      }).format(metric.value);
    case "fraction":
      return new Intl.NumberFormat("en-US", {
        style: "percent",
        maximumFractionDigits: 1,
      }).format(metric.value);
    case "multiple":
      return `${metric.value.toFixed(2)}×`;
    case "days":
      return `${metric.value.toFixed(1)} days`;
  }
}

File: backend/app/schemas/risk_check.py
"""Small, typed Copilot view of the existing risk report, not another engine."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, FiniteFloat

from .confidence import DataConfidence


class CheckMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    key: str
    label: str
    value: FiniteFloat | None = None
    unit: Literal["usd", "fraction", "multiple", "days"]
    horizon: str
    basis: str
    explanation: str
    source_field: str


class CheckFinding(BaseModel):
    key: str
    title: str
    severity: Literal["high", "elevated", "info"]
    explanation: str


class CheckStrategy(BaseModel):
    underlying: str
    expiry: str
    name: str
    leg_count: int
    premium_basis: Literal["entry", "current_mark", "mixed", "unavailable"]
    max_loss: FiniteFloat | None = None
    max_gain: FiniteFloat | None = None
    loss_status: Literal["bounded", "unbounded", "unavailable"]
    gain_status: Literal["bounded", "unbounded", "unavailable"]


class RiskCheck(BaseModel):
    model_config = ConfigDict(extra="forbid")

    portfolio_id: str
    result_id: str
    methodology_version: Literal["risk-check-v1"] = "risk-check-v1"
    computed_at: str
    price_history_as_of: str | None = None
    status: Literal["ready", "limited"]
    summary: str
    metrics: list[CheckMetric] = Field(default_factory=list)
    strategies: list[CheckStrategy] = Field(default_factory=list)
    findings: list[CheckFinding] = Field(default_factory=list, max_length=3)
    limitations: list[str] = Field(default_factory=list)
    data_confidence: DataConfidence | None = None
